server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
         }
     }
 
+     private static Query tryRewriteNumericLongSort(SearchContext searchContext, IndexReader reader) throws IOException {
+        // check sort
+        Sort sort = searchContext.sort().sort;
+        if (Sort.RELEVANCE.equals(sort)) return null;
+        if (Sort.INDEXORDER.equals(sort)) return null;
+        if (sort.getSort().length > 1) return null; // we need only a single sort
+        SortField sortField = sort.getSort()[0];
+
+        // check if this is a field of type Long, that is indexed and has doc values
+        String fieldName = sortField.getField();
+        final MappedFieldType fieldType = searchContext.mapperService().fullName(fieldName);
+        if (fieldType == null) return  null;
+        if (fieldType.typeName() != "long") return null;
+        if (fieldType.indexOptions() == IndexOptions.NONE) return null;
+        if (fieldType.hasDocValues() == false) return null;
+        byte[] minValueBytes =  PointValues.getMinPackedValue(reader, fieldName);
+        byte[] maxValueBytes =  PointValues.getMaxPackedValue(reader, fieldName);
+        if ((maxValueBytes == null) || (minValueBytes == null)) return null; // no values on this shard
+
+        // finally all checks done, query can be rewritten, rewrite
+        long minValue = LongPoint.decodeDimension(minValueBytes, 0);
+        long maxValue = LongPoint.decodeDimension(maxValueBytes, 0);
+        final long origin = (sortField.getReverse()) ? maxValue : minValue;
+        final long pivotDistance = maxValue == minValue ? maxValue/2 : (maxValue - minValue)/2;
```

Comment:
Maybe `minValue == maxValue` is a case when we should return a DocValueFieldExistsQuery since there is no point in computing distances in such a case.